orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright IBM Corp. 2022 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cluster
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/bccsp"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/pkg/errors"
+)
+
+type suspectSet map[string]struct{}
+
+func (s suspectSet) insert(entry string) {
+	s[entry] = struct{}{}
+}
+
+func (s suspectSet) remove(entry string) {
+	delete(s, entry)
+}
+
+func (s suspectSet) has(entry string) bool {
+	_, ok := s[entry]
+	return ok
+}
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+	Close()
+}
+
+type FetcherConfig struct {
+	// TODO: The fields of this struct are evolving
+	Endpoints    []EndpointCriteria
+	FetchTimeout time.Duration
+}
+
+type TimeFunc func() time.Time
+type BlockFetcher struct {
+	MaxPullBlockRetries      uint64
+	BlockSourceFactory       func(c FetcherConfig) BlockSource
+	Config                   FetcherConfig
+	Endpoints                []EndpointCriteria
+	suspects                 suspectSet // a set of bft suspected nodes.
+	Logger                   *flogging.FabricLogger
+	currentEndpoint          EndpointCriteria
+	currentSeq               uint64
+	ShuffleTimeout           time.Duration
+	LastShuffledAt           time.Time
+	MaxByzantineFaults       int64
+	ConfirmByzantineBehavior func(attestations []*common.Block) bool
+	ShuffleTimeoutThrehold   int64 // expressed as %age of FetchTimeOut
+	TimeNow                  TimeFunc
+}
+
+func (bf *BlockFetcher) pullBlockFromSource(seq uint64) *common.Block {
+	bf.currentSeq = seq
+
+	// create a local config
+	config := bf.Config
+	// fill only current endpoint criteria
+	config.Endpoints = []EndpointCriteria{bf.currentEndpoint}
+
+	blockSource := bf.BlockSourceFactory(config)
+	// TODO: Check if the below line is redundant
+	blockSource.UpdateEndpoints([]EndpointCriteria{bf.currentEndpoint})
+	// pull block
+	block := blockSource.PullBlock(seq)
+	// to prevent connection leak, close the source
+	blockSource.Close()
```

Comment:
we should only close it if we shuffle, not each time

Reply:
sure